app/javascript/controllers/tree_controller.js

```diff
@@ -0,0 +1,230 @@
+        parentBranchRowElement.focus()
+        break;
+      case event.key === 'ArrowUp':
+        const previousRowElement = this.previousBranchRowElement(event.target)
+        if(!previousRowElement) break;
+        
+        event.preventDefault()
+        previousRowElement.focus()
+        break;
+      case event.key === 'ArrowDown':
+        const nextRowElement = this.nextBranchRowElement(event.target)
+        if(!nextRowElement) break;
+        
+        event.preventDefault()
+        nextRowElement.focus()
+        break;
+      case event.key === 'Home':
+        event.preventDefault()
+        this.tableTarget.querySelector('tbody > tr').focus()
+        break;
+      case event.key === 'End':
+        event.preventDefault()
+        const rowElements = 
+        this.visibleRows[rowElements.length - 1].focus()
+        break;
+  
+    }
+  }
+
+  isBranch(element) {
+    return element.dataset.treeRole === 'branch'
+  }
+
+  isClosedBranch(element) {
+    return element.dataset.treeRole === 'branch' && element.getAttribute('aria-expanded') !== 'true'
+  }
+
+  isOpenBranch(element) {
+    return element.dataset.treeRole === 'branch' && element.getAttribute('aria-expanded') === 'true'
+  }
+
+  isLeaf(element) {
+    return element.dataset.treeRole === 'leaf'
+  }
+
+  downloadElement(rowElement) {
+    return rowElement.querySelector('a')
+  }
+
+  parentBranchRowElement(rowElement) {
+    let previousElementSibling = rowElement.previousElementSibling
+    while(previousElementSibling) {
+      if(previousElementSibling.getAttribute('aria-level') == rowElement.getAttribute('aria-level') - 1) {
+        return previousElementSibling
+      }
+      previousElementSibling = previousElementSibling.previousElementSibling
+    }
+    return null
+  }
+
+  nextBranchRowElement(rowElement) {
+    let siblingElement = rowElement.nextElementSibling
+    while(siblingElement) {
+      if(this.isVisible(siblingElement)) {
+        return siblingElement
+      }
+      siblingElement = siblingElement.nextElementSibling
+    }
+    return null
+  }
+
+  previousBranchRowElement(rowElement) {
+    let siblingElement = rowElement.previousElementSibling
+    while(siblingElement) {
+      if(this.isVisible(siblingElement)) {
+        return siblingElement
+      }
+      siblingElement = siblingElement.previousElementSibling
+    }
+    return null
+  }
+
+  isVisible(element) {
+    return !this.isHidden(element)
+  }
+
+  isHidden(element) {
+    return element.classList.contains('hidden') || element.classList.contains('search-hidden')
+  }
+
+  updateCount() {
+    const count = this.countTarget.textContent = this.tableTarget.querySelectorAll('tbody > tr[data-tree-role="leaf"]:not(.search-hidden)').length
+    if (count == 1) {
+      this.countTarget.textContent = '1 file'
+    } else {
+      this.countTarget.textContent = `${count} files`
+    }
+  }
+
+  restripe() {
+    this.visibleRows.forEach((rowElement, index) => {
```

Comment:
I believe this can be done in plain css, right?

Reply:
Unfortunately, I don't think CSS striping works in this case since it needs to adjust the striping based on what is visible.

Reply:
`tr:visible:nth-child(odd)` and `[visible]:nth-child(odd)` don't work.

`tr:nth-child(odd)` stripes, but doesn't take into account visibility.

Reply:
sorry, I meant to say hidden rather than visible

Reply:
@jcoyne None of these worked:
```
[visible]:nth-child(odd)  {
  background-color: $color-beige-10;
}

tr[visible]:nth-child(odd)  {
  background-color: $color-beige-10;
}

tr:visible:nth-child(odd)  {
  background-color: $color-beige-10;
}
```

Reply:
@justinlittman yes, you used "visibile", but we want `not(.hidden)`

Reply:
I've tried `not(.hidden)` as well.